deploy/ciao-deploy-client/main.go

```diff
@@ -0,0 +1,474 @@
+	}
+
+	fmt.Printf("Certificate installed in: %s\n", systemCertPath)
+	return nil
+}
+
+func getCertFingerprint(path string) (string, error) {
+	certBytes, err := ioutil.ReadFile(path)
+	if err != nil {
+		return "", errors.Wrap(err, "Error reading certificate from file")
+	}
+
+	certBlock, _ := pem.Decode(certBytes)
+	if certBlock == nil {
+		return "", errors.Wrap(err, "Error decoding PEM data")
+	}
+
+	var cert interface{}
+	if certBlock.Type == "CERTIFICATE REQUEST" {
+		cert, err = x509.ParseCertificateRequest(certBlock.Bytes)
+		if err != nil {
+			return "", errors.Wrap(err, "Error parsing certificate")
+		}
+	} else if certBlock.Type == "CERTIFICATE" {
+		cert, err = x509.ParseCertificate(certBlock.Bytes)
+		if err != nil {
+			return "", errors.Wrap(err, "Error parsing certificate")
+		}
+	} else {
+		return "", errors.New("Unknown PEM type")
+	}
+
+	fp := certs.FingerPrint(cert)
+	return fp, nil
+}
+
+// If the CSR certificate is locally available then return the fingerprint from it, otherwise generate a new CSR and post to the server.
+func getCSRFingerprint() (string, error) {
+	path := csrPath(ssntpRole)
+	if !force {
+		if _, err := os.Stat(path); !os.IsNotExist(err) {
+			fp, err := getCertFingerprint(path)
+			if err != nil {
+				return "", errors.Wrap(err, "Error getting fingerprint")
+			}
+			fmt.Printf("CSR already present. Fingerprint: %s\n", fp)
+			return fp, nil
+		}
+	}
+
+	fp, err := uploadCSR()
+
+	if err != nil {
+		return "", errors.Wrap(err, "Error getting fingerprint")
+	}
+	return fp, nil
+}
+
+// Provides path to role certificate. If not locally available tries to request from the server possibly creating a
+// CSR in the process and submitting that to the server (uses a local CSR if already present.)
+func obtainCert() (string, error) {
+	systemCertPath := path.Join(systemPkiLocation, deploy.CertName(ssntpRole))
+
+	// TODO: Check if certificate installed is valid
+	if !force {
+		if _, err := os.Stat(systemCertPath); !os.IsNotExist(err) {
+			fmt.Printf("Certificate already installed. Skipping download.\n")
+			return systemCertPath, nil
+		}
+	}
+
+	fp, err := getCSRFingerprint()
+	if err != nil {
+		return "", errors.Wrap(err, "error getting CSR fingerprint")
+	}
+
+	err = downloadSignedCert(systemCertPath, fp)
+	if err != nil {
+		return "", errors.Wrap(err, "error downloading signed certificate")
+	}
+
+	// Cleanup saved CSR and private key
+	os.Remove(privPath(ssntpRole))
+	os.Remove(csrPath(ssntpRole))
+
+	return systemCertPath, nil
+}
+
+// Provides path to CA certificate. If not locally available then this is downloaded from the server.
+func obtainCACert() (string, error) {
+	systemCaCertPath := path.Join(systemPkiLocation, "CAcert.pem")
+
+	// TODO: Check if certificate installed is valid
+	if !force {
+		if _, err := os.Stat(systemCaCertPath); !os.IsNotExist(err) {
+			fmt.Printf("CA certificate already installed. Skipping download.\n")
```

Comment:
I'm not sure what should be done in such circumstances, but just wanted to note this.